File: pylibs/brain_metrics/brain_metrics/goal_type.py
"""
goal_type.py — GoalType enum (resolves Child-0 A1 #8).

@paradigm: sql (pure enum, zero LLM)
Justified: GoalType is a discriminated union for metric goal values.
'money' goals carry BIGINT minor units; 'ratio' goals carry INT32 basis
points. This split resolves the Child-0 A1 #8 ambiguity and is the
binding type-contract decision for Child 4's metric engine.

Byte-identity pair with packages/lib-metrics/src/goal-type.ts.
"""

from __future__ import annotations

from enum import StrEnum


class GoalType(StrEnum):
    """Discriminated union for metric goal value types.

    MONEY: goal value is expressed in integer minor units (BIGINT/Int64).
           e.g. a GMV target of ₹100,000 = 10_000_000 paise.
    RATIO: goal value is expressed in basis points (INT32, FLOOR ×10,000).
           e.g. a 25% RTO rate goal = 2500 bp.

    Uses StrEnum (Python 3.11+) so str(GoalType.MONEY) == "money" and
    JSON serialization produces "money" / "ratio" (matching the TS GoalType
    literal union). CF-C2-PRIMITIVE-1 byte-identity pair.
    """

    MONEY = "money"
    RATIO = "ratio"

File: packages/lib-metrics/src/goal-type.ts
// @paradigm: sql
// Resolves Child-0 A1 #8: goalType money|ratio split.
// Design-only this child — the WorkspaceMetricGoal.goalType column migration is in the runbook skeleton, NOT applied.

/**
 * GoalType discriminates between money goals (stored as BIGINT minor units + currency_code)
 * and ratio goals (stored as INT32 basis points).
 *
 * "money" → goalValue is BIGINT minor units (MU), e.g. 100_00000n (₹1,00,000 in paise).
 * "ratio" → goalValue is INT32 basis points, e.g. 2333 (23.33%).
 *
 * This resolves the Child-0 open question A1 #8 (goalType split).
 * The future WorkspaceMetricGoal.goalType enum column migration is documented in the
 * runbook skeleton — it is NOT applied this child (CF-C2-NO-LIVE-1).
 */
export type GoalType = 'money' | 'ratio';

/**
 * Type-safe goalValue discriminated union.
 * "money" carries minorUnits (bigint) + currencyCode.
 * "ratio" carries basisPoints (number, INT32).
 */
export type GoalValue =
  | { readonly goalType: 'money'; readonly minorUnits: bigint; readonly currencyCode: string }
  | { readonly goalType: 'ratio'; readonly basisPoints: number };
